src/core.js

```diff
@@ -221,9 +224,9 @@ const alignWithDOM = function(nodeName, key) {
 
   // Check to see if the node has moved within the parent.
   if (key) {
-    node = keyMap[key];
-    if (node && process.env.NODE_ENV !== 'production') {
-      assertKeyedTagMatches(getData(node).nodeName, nodeName, key);
+    const keyNode = keyMap[key];
+    if (matches(nodeName, key, keyNode)) {
+      node = keyNode;
     }
```

Comment:
I had first done exactly that (calling `removeChild`), but noticed that it threw an error when `keyNode === currentNode`, since `replaceChild` would be called on it right afterwards on line 255, which fails since it wouldn't have a parent anymore.

And in the end, even if it's not the current node, won't this element be removed from the parent anyway via `clearUnvisitedDOM`?

Or would this be an optimization to avoid dealing with this node at all later? If so I could remove the node only when it's not the current one. What do you think?